src/nodes.coffee

```diff
@@ -2089,11 +2105,16 @@ exports.Code = class Code extends Base
             ref = param.asReference o
           else
             if param.value? and not param.assignedInBody
-              ref = new Assign new Value(param.name), param.value
+              ref = new Assign new Value(param.name), param.value, null, param: yes
             else
               ref = param
-          # Add this parameter’s reference to the function scope
-          o.scope.parameter fragmentsToText (if param.value? then param else ref).compileToFragments o
+          # Add this parameter’s reference(s) to the function scope.
+          if param.name instanceof Arr or param.name instanceof Obj
+            # This parameter is destructured.
+            param.name.eachName (prop) ->
```

Comment:
You're right, I forgot `Param` has its own `eachParamName` which is being used [here](https://github.com/jashkenas/coffeescript/pull/4478/files/1e6d910e814275e0c6bd8d781e9d216f9501aa8b..6c814a09780c5c03c7134a7eff442c0512572090#diff-d149b4f007169f06a8ad463f9b2f277bR2035) to check uniqueness.

Some of that iteration could maybe be be centralised (probably later).